src/libraries/System.Memory/tests/Span/Fill.cs

```diff
@@ -147,5 +150,53 @@ public static unsafe void FillNativeBytes()
                 Marshal.FreeHGlobal(new IntPtr(ptr));
             }
         }
+
+        public static IEnumerable<object[]> FillData
+        {
+            get
+            {
+                yield return new object[] { typeof(sbyte), (sbyte)0x20 };
+                yield return new object[] { typeof(byte), (byte)0x20 };
+                yield return new object[] { typeof(bool), true };
+                yield return new object[] { typeof(short), (short)0x1234 };
+                yield return new object[] { typeof(ushort), (ushort)0x1234 };
+                yield return new object[] { typeof(char), 'x' };
+                yield return new object[] { typeof(int), (int)0x12345678 };
+                yield return new object[] { typeof(uint), (uint)0x12345678 };
+                yield return new object[] { typeof(long), (long)0x0123456789abcdef };
+                yield return new object[] { typeof(ulong), (ulong)0x0123456789abcdef };
+                yield return new object[] { typeof(nint), unchecked((nint)0x0123456789abcdef) };
+                yield return new object[] { typeof(nuint), unchecked((nuint)0x0123456789abcdef) };
+                yield return new object[] { typeof(float), 1.0f };
+                yield return new object[] { typeof(double), 1.0d };
+                yield return new object[] { typeof(string), "Hello world!" }; // shouldn't go down SIMD path
+                yield return new object[] { typeof(decimal), 1.0m }; // shouldn't go down SIMD path
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(FillData))]
+        public static void FillWithRecognizedType(Type expectedType, object value)
+        {
+            Assert.IsType(expectedType, value);
+            typeof(SpanTests).GetMethod(nameof(FillWithRecognizedType_RunTest), BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Static)
+                .MakeGenericMethod(expectedType)
+                .Invoke(null, BindingFlags.DoNotWrapExceptions, null, new object[] { value }, null);
+        }
+
+        private static void FillWithRecognizedType_RunTest<T>(T value)
```

Comment:
Maybe I'm misremembering, but doesn't xunit let you put [InlineData] with different types directly onto a generic test method, and it handles doing the relevant reflection/casting for you?

Reply:
Looks like it's not supported. See https://github.com/xunit/xunit/issues/1378.

Reply:
I'm surprised, given we have tests like these:
https://github.com/dotnet/runtime/blob/907fc3e13a53820329890e59d8ff65eb286cca30/src/libraries/System.Collections/tests/Generic/Comparers/Comparer.Tests.cs#L14-L21
https://github.com/dotnet/runtime/blob/907fc3e13a53820329890e59d8ff65eb286cca30/src/libraries/System.Drawing.Common/tests/PenTests.cs#L23-L25
https://github.com/dotnet/runtime/blob/907fc3e13a53820329890e59d8ff65eb286cca30/src/libraries/System.Linq/tests/GroupByTests.DebuggerAttributes.cs#L13-L15

Reply:
Huh, for some reason xunit always kept passing _T = object_ in for this unit test when I tried doing this. I'll try tweaking the signature to see if this fixes things.

Reply:
Setting up the theory to support this ended up making the code look bad. I simplified a bit to put all the data inline and have it delegate to a common test helper. This avoids the reflection call and looks cleaner overall.